chore(auditor): refactor spends fetch to be background progressing

joshuef opened the pull request:
This pull request mainly focuses on improving the `sn_auditor` package, particularly the `dag_db.rs` and `main.rs` files. The changes involve adding new libraries, modifying existing methods, and introducing new methods for better DAG collection and handling. The changes are aimed at enhancing the functionality of the `SpendDagDb` struct and improving the overall performance of the software.

The most important changes include:

* Addition of new libraries and modification of existing imports:
  * `tokio = { version = "1.32.0", features = [` in file `sn_auditor/Cargo.toml`: Added `itertools = "0.13.0"` and `futures = "0.3.30"` libraries.
  * `use bls::SecretKey;` in file `sn_auditor/src/dag_db.rs`: Added `use futures::future::join_all;` and modified the `use sn_client::transfers::` import.
  * `use std::time::{SystemTime, UNIX_EPOCH};` in file `sn_auditor/src/dag_db.rs`: Added `use std::time::{Duration, Instant};`.

* Changes to the `SpendDagDb` struct in `sn_auditor/src/dag_db.rs`:
  * Modified the access modifier of `beta_participants` from `pub(crate)` to private and added the `REATTEMPT_INTERVAL` constant.
  * Removed the `update` method and modified the `track_new_beta_participants` and `init_reward_forward_tracking` methods to use `BTreeSet<String>` instead of `Vec<String>`. [[1]](diffhunk://#diff-fc4319f01fee1895543b912d85d9455555b052ecc12a54539497148590b3dc48L183-L232) [[2]](diffhunk://#diff-fc4319f01fee1895543b912d85d9455555b052ecc12a54539497148590b3dc48L266-R222) [[3]](diffhunk://#diff-fc4319f01fee1895543b912d85d9455555b052ecc12a54539497148590b3dc48R252-R438)
  * Introduced new methods `backup_rewards`, `dag_crawling_from_genesis`, and `process_spend`.

* Changes to `sn_auditor/src/main.rs`:
  * Modified the `initialize_background_spend_dag_collection` function to use `BTreeSet<String>` instead of `Vec<String>` for `beta_participants` and adjusted the logic for updating the DAG. [[1]](diffhunk://#diff-41fbce10df15210e82c92ee34bc0329d4eacabd75c03a2bd9f2cf59b3a208828R19-R32) [[2]](diffhunk://#diff-41fbce10df15210e82c92ee34bc0329d4eacabd75c03a2bd9f2cf59b3a208828L190-R190) [[3]](diffhunk://#diff-41fbce10df15210e82c92ee34bc0329d4eacabd75c03a2bd9f2cf59b3a208828L204-R214) [[4]](diffhunk://#diff-41fbce10df15210e82c92ee34bc0329d4eacabd75c03a2bd9f2cf59b3a208828L244-R245)
  * Adjusted the `load_and_update_beta_participants` function to use `BTreeSet<String>` instead of `Vec<String>`. [[1]](diffhunk://#diff-41fbce10df15210e82c92ee34bc0329d4eacabd75c03a2bd9f2cf59b3a208828L314-R317) [[2]](diffhunk://#diff-41fbce10df15210e82c92ee34bc0329d4eacabd75c03a2bd9f2cf59b3a208828L348-R339)

* Changes to `sn_auditor/src/routes.rs` and `sn_client/src/api.rs`:
  * Modified the `add_participant` function in `sn_auditor/src/routes.rs` to use `BTreeSet<String>` instead of `Vec<String>`.
  * Adjusted the `crawl_spend_from_network` function in `sn_client/src/api.rs` to return a tuple containing `SpendAddress` and `Result<SignedSpend>`. [[1]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L947-R952) [[2]](diffhunk://#diff-6aa8955102cfa98bc83700a7a1860ae21ace8bc5545df1ef0e4d6388defaab97L957-R963)

joshuef commented:
Nice one @RolandSherwin ! Good work here 💪 

So you've got all the separation in here, but I think we can push it a bit further (and I think without too much difficulty, I hope). Since we have the channels and the dysf operation, can we pull it out of `MyNode` altogether?

I think the `DysfunctionTracking` could be kicked off at the `FlowCtrl` layer, perhaps? With the `Sender` passed into the node? (At least as a first pass... I know we have these `log_comm_issue` fns that take place where we can't always return Cmds... So it means for now the `Sender` needs to be available for those APIs in `MyNode`.)

----

Down the line, I'd like to verify that our pattern for AeDysfunction etc makes sense... I'm not 100% sure at the moment. But that's not one for this PR

joshuef reviewed (changes requested):
Lets pull the whole Dysf out of Mynode, as it doesn't need to be strictly tied to the node state.

joshuef commented on sn_node/src/node/flow_ctrl/dysfunction.rs:
```diff
@@ -0,0 +1,103 @@
+        let _ = tokio::task::spawn(async move {
+            debug!("[NODE READ]: flowctrl start dysfunction detection");
+            let mut dysfunction = DysfunctionDetection::new(
+                node.read()
```
can we pass in context here instead?

joshuef commented on sn_node/src/node/flow_ctrl/dysfunction.rs:
```diff
@@ -0,0 +1,103 @@
+    pub(crate) async fn get_dysfunctional_node_names(&mut self) -> BTreeSet<XorName> {
+        // send a DysCmd asking for the dysfunctional nodes
+        debug!("[NODE READ]: flowctrl send DysCmds to read dysfunctional nodes");
+        if let Err(error) = self
```
we shouldn't need to reach into the `node` for this. `Sender` is `clone`, so we can keep an instance of that on `FlowCtrl`

joshuef commented:
bors r+

maqi committed:
chore(auditor): refactor spends fetch to be background progressing

maqi committed:
chore(auditor): do not track royalty and payment_forward utxo

joshuef committed:
feat(audit): use BTreeSet for participants

To avoid duplication

joshuef committed:
feat(audit): parallelise queries

joshuef committed:
HACK:fix protocol version

joshuef committed:
feat(audit): enable url decoding

fix: #1778

maqi commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,156 @@ impl SpendDagDb {
+
+                if let Some(spend) = spend {
+                    Self::process_spend(
+                        // &mut utxo_addresses,
```
can now be removed.
same to the other places

maqi commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,156 @@ impl SpendDagDb {
+                };
+
+                if let Some(spend) = spend {
+                    Self::process_spend(
```
if make function to be a `member function`, i.e. call like `self.process_spend()`
then we can avoid pass down those `encryption_sk`, `beta_participants`, `beta_participants` separately ?

maqi commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,156 @@ impl SpendDagDb {
+        let mut new_utxos: Vec<SpendAddress> = vec![];
+
+        // Filter out royalty outputs
+        let mut royalty_pubkeys = BTreeSet::new();
```
```
let royalty_pubkeys: BTreeSet<_> = spend.spend.network_royalties.iter().map(|derivation_idx| {
            NETWORK_ROYALTIES_PK.new_unique_pubkey(derivation_idx)
        }).collect();
```

maqi commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,156 @@ impl SpendDagDb {
+            royalty_pubkeys.insert(unique_pubkey);
+        }
+
+        for output in spend.spend.spent_tx.outputs.iter() {
```
```
let new_utxos: Vec<_> = spend.spend.spent_tx.outputs.iter().filter_map(|output| {
            if !royalty_pubkeys.contains(&output.unique_pubkey) {
                Some(SpendAddress::from_unique_pubkey(&output.unique_pubkey))
            } else {
                None
            }
        }).collect();
```

maqi commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,156 @@ impl SpendDagDb {
+                .insert((addr, amount));
+        }
+
+        // for new_utxo in new_utxos {
```
can now be removed

maqi commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,156 @@ impl SpendDagDb {
+                    .collect();
+                addresses.extend(re_attempt_address);
+                utxo_addresses.retain(|_, time_stamp| *time_stamp > Instant::now());
+            }
```
better insert a sleep if addresses is still empty
in case those utxo_addresses are ALL not timed_out yet, 
```
if addresses.is_empty() {
    Sleep(REATTEMPT_INTERVAL);
    continue;
}
```

maqi commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,156 @@ impl SpendDagDb {
+        let mut self_payments = forwarded_payments.write().unwrap();
+
+        if let Some(user_name) = beta_participants_read.get(&user_name_hash) {
+            eprintln!("Got forwarded reward from {user_name} of {amount} at {addr:?}");
```
this can now be changed back to `info` or `trace` ?
same to the other

joshuef committed:
chore(audit): refactor and cleanup code

Address feedback, remove comments cleaner set collection and small await if nothing to scan

grumbach commented on sn_auditor/src/main.rs:
```diff
@@ -241,20 +233,16 @@ async fn initialize_background_spend_dag_collection(
         loop {
-            println!("Updating DAG...");
-            let _ = d
-                .update()
```
Removing the update here means we dump the same DAG over and over at each iteration.

grumbach commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,147 @@ impl SpendDagDb {
+                let spend = match result {
+                    Ok(spend) => Some(spend),
+                    Err(_) => {
+                        let _ = utxo_addresses.insert(address, Instant::now() + REATTEMPT_INTERVAL);
```
This is a very smart idea and could be used in other areas where we do DAG crawling from UTXOs

grumbach reviewed (changes requested):
Will take over from here and grab the nice optimisations to another PR so we don't need to split

grumbach commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,147 @@ impl SpendDagDb {
+
+    /// Helper function to process each spend and update relevant data structures.
+    async fn process_spend(&self, addresses: &mut BTreeSet<SpendAddress>, spend: SignedSpend) {
+        // Filter out royalty outputs
+        let royalty_pubkeys: BTreeSet<_> = spend
```
This makes sense for beta as long as we assume that the Foundation's money won't move from its wallet EVER. 
As soon as the money is back into the economy this assumption would leave whole DAG branches in the dark. 
For now though, this is an amazing optimization as half of the data related UTXOs are indeed royalties! 
Good one, but let's just make sure we don't forget!

grumbach commented on sn_auditor/src/dag_db.rs:
```diff
@@ -382,6 +287,147 @@ impl SpendDagDb {
+            for (address, result) in spends {
+                let spend = match result {
+                    Ok(spend) => Some(spend),
+                    Err(_) => {
```
Error management is omitted here: all errors here are treated similarly, this includes `RecordNotFound` where the retry after `REATTEMPT_INTERVAL` behavior is indeed wanted, but we might not want this for other errors like `DoubleSpend` or others. We don't want to end up in a retry forever loop on error.

joshuef commented:
Closing as #1785 replaces this.
